File: src/openbook/core/models/mixins/i18n.py
# OpenBook: Interactive Online Textbooks - Server
# © 2025 Dennis Schulmeister-Zimolong <dennis@wpvs.de>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.

from typing                     import Optional
from django.conf                import settings
from django.db                  import models
from django.db.models           import Q
from django.utils.translation   import gettext_lazy as _, get_language

def LanguageField():
    """
    Return a special model field for language codes.

    Technically this is a simple foreign key to the ``Language`` model of the core app.
    """
    return models.ForeignKey("openbook_core.Language", on_delete=models.CASCADE)

class TranslatableMixin(models.Model):
    """
    Provide a mixin for translation models with translated texts for a parent model.

    Given a model named ``Example``, translations can be added like this::

        class Example_T(models.Model, TranslatableMixin):
            parent = models.ForeignKey(Example, on_delete=models.CASCADE, related_name="translations")

            text1  = models.CharField(verbose_name=_("Some Text 1"), max_length=255, null=False, blank=False)
            text2  = models.CharField(verbose_name=_("Some Text 2"), max_length=255, null=False, blank=False)
            text3  = models.CharField(verbose_name=_("Some Text 3"), max_length=255, null=False, blank=False)

            class Meta(TranslatableMixin.Meta):
                verbose_name        = _("My Model: Translation")
                verbose_name_plural = _("My Model: Translations")

    The translation model simply needs a foreign key to the parent model (usually called
    ``parent`` with related name ``translations``) and char fields for the translatable texts.

    This class can be combined with the ``UUIDMixin``, if the parent class used it, too.
    """
    language = LanguageField()

    class Meta:
        abstract = True
        ordering = ("parent", "language")
        indexes  = (models.Index(fields=("parent", "language")),)

    def __str__(self):
        return self.language.name

def get_translations(object: models.Model, language: str = "",
                     attr_id: str = "id", attr_translations: str = "translations",
                     attr_t_parent: str = "parent", attr_t_language: str = "language") -> Optional[models.Model]:
    """
    Get translations of a model with translations.

    By default translations are stored in a second model, that installs a
    ``translations`` (``attr_translations``) related attribute. The text model usually
    has the properties ``parent`` (``attr_t_parent``) pointing to the original model and
    ``language`` (``attr_t_language``) with the language code.

    Tries to find translations for the given language (default: language of the current thread)
    or the ``LANGUAGE_CODE`` setting as fallback, if different.

    :returns: The best found translation or ``None``, if none exists.
    """
    if not language:
        language = get_language()

    id = getattr(object, attr_id)

    default_language = settings.LANGUAGE_CODE or ""
    default_kwargs   = {attr_t_parent: id, attr_t_language: default_language}
    language_kwargs  = {attr_t_parent: id, attr_t_language: language}

    translations = getattr(object, attr_translations)

    if default_kwargs == language_kwargs:
        results = translations.filter(**language_kwargs);
    else:
        results = translations.filter(Q(**language_kwargs) | Q(**default_kwargs))

    if results.count() == 0:
        return None
    elif results.count() == 1:
        return results.first()
    else:
        return results.get(**language_kwargs)

class TranslationMissing(Exception):
    """
    Signal that a translation for something is missing.

    Note that the ``get_translations()`` function does not throw this exception but
    rather returns ``None``.
    """
    pass

File: src/frontend/app/src/stores/i18n.ts
/*
 * OpenBook: Interactive Online Textbooks
 * © 2024 Dennis Schulmeister-Zimolong <dennis@wpvs.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 */

import type {I18N}         from "../i18n/index.js";
import type {LanguageCode} from "../i18n/index.js";

import {fallbackLanguage}  from "../i18n/index.js";
import {defaultLanguage}   from "../i18n/index.js";
import {ReadableStore}     from "../utils/store.js";
import {WritableStore}     from "../utils/store.js";

export {languages}         from "../i18n/index.js";
export {defaultLanguage}   from "../i18n/index.js";
export {fallbackLanguage}  from "../i18n/index.js";

const i18nDefault = await createTranslations(defaultLanguage);

class I18nStore extends ReadableStore<I18N> {
    constructor() {
        super(i18nDefault);

        language.subscribe(async newLanguage => {
            this.set(await createTranslations(newLanguage));
        });
    }
}

/**
 * The current UI language. To change the language simple assign a new value.
 * The UI will rerender accordingly.
 */
export const language = new WritableStore<LanguageCode>(defaultLanguage);

/**
 * Message catalogue with all translations of the currently active language.
 * This is just a deeply structured key/value object, that can be directly
 * accessed with `$i18n.someKey` in the UI components.
 */
export const i18n = new I18nStore();

/**
 * Utility function to replace placeholders in the form of `$key$` in the given
 * text with the property of the object given in the second parameter.
 *
 * @param text Original text
 * @param values Key/values to insert
 * @return Text with replaced placeholders
 */
export function _(text: string, values: any): string {
    for (let key of Object.keys(values) || []) {
        text = text.replaceAll(`$${key}$`, values[key]);
    }

    return text;
}

/**
 * Create a new message catalogue from the given langauge and the fallback language.
 *
 * @param newLanguage Language code
 * @returns New message catalogue
 */
async function createTranslations(newLanguage: string): Promise<I18N> {
    let i18n = await import(`../i18n/lang/${fallbackLanguage}.ts`);
    let translations = deepCopy({}, i18n.default);

    if (newLanguage !== fallbackLanguage) {
        let i18n = await import(`../i18n/lang/${newLanguage}.ts`);
        deepCopy(translations, i18n.default);
    }

    return translations;
}

/**
 * Deep copy of an object with translatable texts, similar to `Object.assign()`.
 * Note that the object properties must be strings or nested translation objects.
 * If a property is an object, the function will copy its content by recursively
 * calling itself. All other values are copied via a simple assignment.
 *
 * Properties of the target object that are missing in the source object remain
 * unmodified.
 *
 * @param target Target object
 * @param source Source object
 * @returns Target object
 */
function deepCopy(target: any, source: any): any {
    for (let key of Object.keys(source)) {
        let value = source[key];

        if (typeof value === "object") {
            target[key] = {};
            deepCopy(target[key], source[key]);
        } else {
            target[key] = source[key];
        }
    }

    return target;
}
